### ISSUE TITLE
'Total available for each user' in cart rule not considered

### ISSUE THREAD

**Issue by idnovate** [PrestaShop/PrestaShop#15751]
**Describe the bug**
When you offer a product with price 0 as a free gift any record is inserted at table `ps_order_cart_rule`, so the customer can use it another time

**To Reproduce**

1. Set a product with price 0
![image](https://user-images.githubusercontent.com/15104724/65875677-0d514c80-e388-11e9-939d-605dbe2f9ae8.png)
1. Create a cart rule limited to 1 use per customer and offer this product as free gift:
![image](https://user-images.githubusercontent.com/15104724/65875720-222de000-e388-11e9-86c6-d27a94fda9a7.png)
![image](https://user-images.githubusercontent.com/15104724/65875744-2fe36580-e388-11e9-9747-90f03e398f9c.png)

1. Create an order. Cart rule is added correctly.
1. Create a second order. **Cart rule is added but it shouldn't**.

**Additional information**
PrestaShop version: 1.7.6.1

**Comment by idnovate** [comment]
Record at table `ps_order_cart_rule` is inserted at function `validateOrder()` from `PaymentModule` class, at this line:

```
$cart_rules_list = $this->createOrderCartRules(
    $order,
    $this->context->cart,
    $order_list,
    $total_reduction_value_ti,
    $total_reduction_value_tex,
    $id_order_state
);
```

Function `createOrderCartRules()` uses `$cartRulesData = $calculator->getCartRulesData();`.

Function `getCartRulesData()` is at `\src\Core\Cart\CartRuleCalculator.php` and uses `cartRules` information, which is filled at `applyCartRule()`.

Function `applyCartRule()` execute:
```
// Free gift
if ((int) $cartRule->gift_product) {
    foreach ($this->cartRows as $cartRow) {
        $product = $cartRow->getRowData();
        if ($product['id_product'] == $cartRule->gift_product
            && ($product['id_product_attribute'] == $cartRule->gift_product_attribute
                || !(int) $cartRule->gift_product_attribute)
        ) {
            $cartRuleData->addDiscountApplied($cartRow->getFinalUnitPrice());
            $cartRow->applyFlatDiscount($cartRow->getFinalUnitPrice());
        }
    }
}
```

and retrieves product price, which is in this case is 0.

If we return to first function `createOrderCartRules()`, it executes:

```
foreach ($cartRulesData as $cartRuleData) {
    $cartRule = $cartRuleData->getCartRule();
    // Here we need to get actual values from cart calculator
    $values = array(
        'tax_incl' => $cartRuleData->getDiscountApplied()->getTaxIncluded(),
        'tax_excl' => $cartRuleData->getDiscountApplied()->getTaxExcluded(),
    );

    // If the reduction is not applicable to this order, then continue with the next one
    if (!$values['tax_excl']) {
        continue;
    }
    ...
```

So `$values['tax_excl']` is 0 and record in table `ps_order_cart_rule` is not inserted.

**Comment by khouloudbelguith** [comment]
Hi @idnovate,

Thanks for your report.
I manage to reproduce the issue with PS1.7.5.2 & PS1.7.6.1.
I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!
Thanks!

**Comment by micka-fdz** [comment]
Hello, any news about this issue?

**Comment by khouloudbelguith** [comment]
@micka-fdz, sorry not yet.
But PrestaShop is an open source project, so it can be solved before if someone submits a pull request to solve it.

Thanks!

**Comment by florine2623** [comment]
This issue is fixed in PS 9.0.0.

https://github.com/user-attachments/assets/85c1d31c-8f4b-461e-a902-3368767d0ce6


I'm closing the issue to avoid clogging the backlog.
